### DISCUSSION TITLE
undefined symbol: _Z9ConstantsRN8nanobind7module_E

### DISCUSSION THREAD

**Discussion by nyckmaia** [wjakob/nanobind#646] [Q&A] [answered]
```
// constants.h

#ifndef CONSTANTS_H
#define CONSTANTS_H

#include <array>
#include <map>
#include <string>
#include <vector>

extern const float a;
extern const float b;
extern const std::vector<float> c;
extern const std::map<int, std::pair<std::string, std::string>> d;

enum class MyEnumClass {
    A,
    B,
    C
};
```

```
// constants.cpp

#include "myproject/constants.h"

const float a = 1.0f;
const float b = 2.0f;
const std::vector<float> c = 3.0f;
const std::map<int, std::pair<std::string, std::string>> d = ...;
```
Nanobind wrapper
--------------------

```
// py_constants.cpp

#include <nanobind/nanobind.h>

#include "maiacore/constants.h"

namespace nb = nanobind;

void Constants(const nb::module_ &m) {
    nb::enum_<MyEnum>(m, "MyEnum")
        .value("A", MyEnum::A)
        .value("B", MyEnum::B);
}

```

```
// myproject.cpp
namespace nb = nanobind;

void Constants(nb::module_ &);
void OtherClass(nb::module_ &);

NB_MODULE(myproject, m) {
    m.doc() = "...";

    Constants(m);
    OtherClass(m);
    ...
}
```

CMakLists.txt
--------------
```
nanobind_add_module(${PROJECT_NAME}
        STABLE_ABI
        NB_STATIC
        ${myproject_src} ${nanobind_wrapper_src}
    )
```
It compiles ok without any error or warning, including the `constants.cpp` file and `py_constants.cpp` file (wrapper).
So, I enter inside the `mymodule.so` directory and tried to import it.

Python Error
--------------
```
Python 3.8.10 (default, Jun 22 2022, 20:18:18)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import myproject as my
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /mnt/c/Users/nyck/Desktop/maialib/build/Linux/module/myproject.cpython-38-x86_64-linux-gnu.so: undefined symbol: _Z9ConstantsRN8nanobind7module_E
>>>
```
Help
-----
Can you help me to fix this Python import?

My System
-------------
- Ubuntu 22 LTS
- C++ Compiler: gcc 9.3.0
- C++ 17
- nanobind 2.0.0

**Comment by nyckmaia** [comment]
UPDATE 01
------------

I found a mistake.
In the file `py_constants.cpp` I use the keyword `const` in the line: `void Constants(const nb::module_ &m) {`.
But in the file `myproject.cpp` I don't use `const`, as you can see: `void Constants(nb::module_ &);`

So, I remove the `const` keyword.
It compiles without any error or warning and genarated the Python module file.
Now, when I tried to import it, I got a different error: `bad_cast`

Env: Linux Ubuntu + GCC 9.3 + Python 3.8.10
--------------------------------------------------
```
Python 3.8.10 (default, Jun 22 2022, 20:18:18)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import maiacore as mc
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: std::bad_cast
```

Env: Windows 11 + Clang 18.3 + Python 3.10.6
----------------------------------------------------
```
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import maiacore as mc
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: bad cast
```
Help
-----

1. How can I debug this `bad_cast` runtime import error to discovery the bug in my code?
2. I have some C++ methods that uses `nlohmann::json` type. Is there any problem with that? Do I have to convert the `nlohmann::json` return type below to a `nb::dict` too?
Exemple:
```
cls.def(
    "myMethod",
    [](A& self, const nb::dict& config) -> nlohmann::json {
        // Python dict type to Json as std::string
        std::string jsonString = nb::cast<std::string>(nb::str(config));

        // JSON as string to nlohmann::json type
        nlohmann::json jsonObj = nlohmann::json::parse(jsonString);

        return self.myMethod(jsonObj);
    },
    nb::arg("config") = nb::dict());
```
3. Is there a better way to convert `nb::dict` to `nlohmann::json` and vice versa?

**Comment by wjakob** [comment, marked as answer]
1. Set a breakpoint on __cxa_throw (this only works on Linux/macOS, not sure how to do it in Windows)
2. AFAIK there aren't yet nanobind bindings for this json library. You'll have to make your own.
3. See point 2. I suggest that you carefully study the `nlohmann_json` pybind11 bindings and convert them to nanobind. Such a conversoin may be useful to others as well. I don't really use JSON and therefore haven't had the time to look into this myself.
